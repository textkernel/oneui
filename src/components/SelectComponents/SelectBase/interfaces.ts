import { GetItemPropsOptions, GetToggleButtonPropsOptions } from 'downshift';

export interface BasicSelectProps<S> extends React.HTMLAttributes<HTMLDivElement> {
    /**
     * an array of objects or strings that will be used to render the suggestions list.
     * if you want to disabled some of the items, this props needs to be an array of objects,
     * where the object has a property "disabled" set true as needed.
     */
    suggestions: S[];
    /** suggestionToString(suggestion) should return a string to be displayed in the UI. e.g.: suggestion => suggestion.name */
    suggestionToString: (suggestions?: S | null) => string;
    /** render function for suggestion list item. If undefined, suggestionToString will be used. */
    suggestionItemRenderer?: (suggestions: S | null, index: number, array: S[]) => ReactNode;
    /** root wrapper ref */
    rootRef?: React.RefObject<HTMLDivElement>;
    /** suggestions list ref */
    listRef?: React.RefObject<HTMLUListElement>;
    /** defines if the component is disabled */
    disabled?: boolean;
    /** a class to be applied to the top level div */
    className?: string;
    /** onFocus() is called when the component is focused */
    onFocus?: () => void;
    /** onBlur() is called when the component is blurred */
    onBlur?: () => void;
    /** onSelectionAdd() called when a suggestion is selected */
    onSelectionAdd: (item: S) => void;
}

export interface SelectInputFieldProps {
    /** input field ref */
    inputRef?: React.RefObject<HTMLElement | undefined>;
    /** onInputValueChange(inputValue) called when the input values is changed. Can be used to implement the component as controlled component */
    onInputValueChange?: (value: string) => void;
    /** clean up input value after selected item */
    clearInputAfterSelection?: boolean;
    /** Initial input value should be shown if input gets focused */
    initInputValue?: string;
}

export interface SelectClearButtonProps {
    /** to be shown as clear button title */
    clearTitle?: string;
    /** function to be called if FieldWrapper clear button is clicked */
    onClearAllSelected?: () => void;
    /** show Clear button on hover even if there are no selectedSuggestions passed */
    showClearButton?: boolean;
}

export type FocusedRendererHelpers<S> = (helpers: {
    getInputProps: (options: GetItemPropsOptions<S>) => object;
    getToggleButtonProps: (options: GetToggleButtonPropsOptions) => object;
    onBlur: () => void;
    onFocus: (callback: () => void) => void;
    inputValue: string;
    highlightedIndex: number | null;
}) => ReactNode;

export type BlurredRendererHelpers<S> = (helpers: {
    getInputProps: (options: GetItemPropsOptions<S>) => object;
    getToggleButtonProps: (options: GetToggleButtonPropsOptions) => object;
    onFocus: (callback: () => void) => void;
    onBlur: () => void;
}) => ReactNode;

export type ListRendererHelper<S> = (props: {
    /** An array of objects that will be used to render the suggestions list. */
    suggestions: S[];
    /** suggestionToString(suggestion) should return a string to be displayed in the UI. e.g.: suggestion => suggestion.name */
    suggestionToString: (suggestions: S) => string;
    /** a function which gets props for the item in the list */
    getItemProps: (options: GetItemPropsOptions<S>) => object;
    /** index of the item from the list to be highlighted */
    highlightedIndex: number | null;
    /** input field value to be highlighted in the item from the list */
    inputValue: string;
}) => ReactNode;

export interface Props<S>
    extends BasicSelectProps<S>,
        SelectClearButtonProps,
        SelectInputFieldProps {
    /** defines if the suggestion list should be collapsed once an item is selected */
    keepExpandedAfterSelection?: boolean;
    /** will be called when list of suggestions should be rendered */
    listRenderer: ListRendererHelper<S>;
    /** a function that renders the top part of the component when it is focused  */
    focusedRenderer: FocusedRendererHelpers<S>;
    /** a function that renders the top part of the component when it is blurred  */
    blurredRenderer: BlurredRendererHelpers<S>;
    /** highlighting for first item will be added if input is empty */
    highlightOnEmptyInput?: boolean;
    /** select highlighted item when blurring out of the component */
    selectOnTab?: boolean;
    /** show dropdown icon */
    showArrow?: boolean;
}
