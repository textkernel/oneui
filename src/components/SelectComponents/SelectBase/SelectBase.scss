.SelectBase {
    position: relative;
    min-height: 34px;

    &__main {
        position: relative;
        box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.1);
        background-color: var(--color-background);
        transition-property: box-shadow, transform, border-color;
        transition-duration: var(--transition-duration);
        border-radius: var(--border-radius);

        &--focused {
            position: absolute;
            z-index: 400;
            width: 100%;
        }
    }

    &__field {
        padding: 0;
    }

    &__spacedElem {
        margin: var(--spacing-normal);
    }

    &__list--focused {
        max-height: 300px;
        overflow: auto;
        border-top: 1px solid var(--border-color-light);
        margin-top: -1px;
    }

    &__loaderContainer {
        width: 250px;
    }
}
