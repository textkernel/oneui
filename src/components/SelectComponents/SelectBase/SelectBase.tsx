import * as React from 'react';
import Downshift from 'downshift';
import { bem } from '../../../utils/bem';
import { useBrowserTabVisibilityChange } from '../../../hooks';
import { FieldWrapper } from '../../FieldWrapper';
import { List } from '../../List';
import { Props } from './interfaces';
import styles from './SelectBase.scss';

const { block, elem } = bem('SelectBase', styles);

export function SelectBase<S>(props: Props<S>) {
    const {
        suggestions,
        suggestionToString,
        clearTitle,
        showClearButton,
        selectOnTab,
        disabled,
        onFocus,
        onBlur,
        onSelectionAdd,
        onInputValueChange,
        onClearAllSelected,
        inputRef: inputRefFromProps,
        rootRef: rootRefFromProps,
        listRef: listRefFromProps,
        listRenderer,
        focusedRenderer,
        blurredRenderer,
        keepExpandedAfterSelection,
        clearInputAfterSelection,
        highlightOnEmptyInput,
        initInputValue,
        ...rest
    } = props;

    const [inputRef, setInputRef] = React.useState(
        inputRefFromProps || React.createRef<HTMLInputElement>()
    );
    const [rootRef, setRootRef] = React.useState(
        rootRefFromProps || React.createRef<HTMLDivElement>()
    );
    const [listRef, setListRef] = React.useState(
        listRefFromProps || React.createRef<HTMLUListElement>()
    );

    const [inputValue, setInputValue] = React.useState(initInputValue || '');
    const [inputValueRecall, setInputValueRecall] = React.useState('');
    const [focused, setFocused] = React.useState(false);
    const isBrowserTabVisible = useBrowserTabVisibilityChange();

    // focus input field if component is focused
    React.useEffect(() => {
        if (focused) {
            inputRef.current?.focus();
        }
    }, [focused, inputRef]);

    // update inputRef from props if changed
    React.useEffect(() => {
        if (inputRefFromProps) {
            setInputRef(inputRefFromProps);
        }
    }, [inputRefFromProps]);

    // update listRef from props if changed
    React.useEffect(() => {
        if (listRefFromProps) {
            setListRef(listRefFromProps);
        }
    }, [listRefFromProps]);

    // update rootRef from props if changed
    React.useEffect(() => {
        if (rootRefFromProps) {
            setRootRef(rootRefFromProps);
        }
    }, [rootRefFromProps]);

    React.useEffect(() => {
        setInputValue(initInputValue || '');
    }, [setInputValue, initInputValue]);

    const handleBlur = () => {
        setFocused(false);
        if (isBrowserTabVisible) {
            setInputValue(initInputValue || '');
            setInputValueRecall(initInputValue || '');
            if (focused) {
                onBlur?.();
            }
        }
    };

    const handleChange = (selectedItem, downshift) => {
        const { clearSelection, openMenu } = downshift;

        clearSelection();

        if (selectedItem) {
            onSelectionAdd(selectedItem);
        }

        if (clearInputAfterSelection) {
            setInputValue('');
            onInputValueChange?.('');
        }

        if (!keepExpandedAfterSelection) {
            handleBlur();
        } else {
            openMenu();
        }
    };

    const handleInputValueChange = (inputVal, changes) => {
        if (changes.type === Downshift.stateChangeTypes.changeInput) {
            setInputValue(inputVal);
            setInputValueRecall(inputVal);
            if (listRef.current) {
                listRef.current.scrollTop = 0;
            }

            onInputValueChange?.(inputVal);
        }
    };

    const handleClearSelectedSuggestions = (e) => {
        e.stopPropagation();
        onClearAllSelected?.();
    };

    const focus = (openMenu) => {
        openMenu();
        setFocused(true);
    };

    const handleWrapperClick = (openMenu) => () => {
        if (!focused && !disabled) {
            focus(openMenu);
        }
    };

    const handleInputOnFocus = (openMenu) => () => {
        if (!focused) {
            focus(openMenu);
        }

        if (isBrowserTabVisible) {
            onFocus?.();
        }
    };

    const stateUpdater = (change, state) => {
        switch (change.type) {
            case Downshift.stateChangeTypes.blurInput:
                handleBlur();
                break;
            case Downshift.stateChangeTypes.clickButton:
                setFocused(change.isOpen);
                break;
            case Downshift.stateChangeTypes.changeInput:
                if (state.isOpen !== focused) {
                    setFocused(state.isOpen);
                }
                break;
            default:
                break;
        }
    };

    const stateReducer = (state, newChanges) => {
        const changes = {
            ...newChanges,
            isEscapeAction: false,
        };

        switch (newChanges.type) {
            case Downshift.stateChangeTypes.changeInput:
                return {
                    ...changes,
                    highlightedIndex: highlightOnEmptyInput || newChanges.inputValue ? 0 : -1,
                };
            case Downshift.stateChangeTypes.clickItem:
            case Downshift.stateChangeTypes.keyDownEnter:
                return {
                    ...changes,
                    highlightedIndex: state.highlightedIndex,
                    isOpen: keepExpandedAfterSelection,
                };
            case Downshift.stateChangeTypes.keyDownEscape:
                return {
                    ...changes,
                    isOpen: false,
                    isEscapeAction: true,
                };
            case Downshift.stateChangeTypes.blurInput:
                if (selectOnTab && !state.isEscapeAction && isBrowserTabVisible) {
                    return {
                        ...changes,
                        selectedItem: suggestions[state.highlightedIndex],
                        isOpen: false,
                    };
                }
                return changes;
            default:
                return changes;
        }
    };

    const stateAndProps = { ...props, focused };

    const getInputPropsWithUpdatedRef = (getInputProps) => (inputProps) => {
        return inputRefFromProps
            ? getInputProps(inputProps)
            : getInputProps({ ...inputProps, ref: inputRef });
    };

    return (
        <div {...rest} ref={rootRef} {...block(stateAndProps)}>
            <Downshift
                onChange={handleChange}
                itemToString={suggestionToString}
                onOuterClick={handleBlur}
                stateReducer={stateReducer}
                onStateChange={stateUpdater}
                onInputValueChange={handleInputValueChange}
                inputValue={inputValue}
                defaultHighlightedIndex={highlightOnEmptyInput ? 0 : -1}
            >
                {({
                    getInputProps,
                    getMenuProps,
                    getItemProps,
                    getToggleButtonProps,
                    highlightedIndex,
                    openMenu,
                }) => (
                    <div {...elem('main', stateAndProps)}>
                        <FieldWrapper
                            clearLabel={clearTitle}
                            onClear={handleClearSelectedSuggestions}
                            showClearButton={!focused && showClearButton}
                            isFocused={focused && !disabled}
                            disabled={disabled}
                            onClick={handleWrapperClick(openMenu)}
                            {...elem('field', stateAndProps)}
                        >
                            {focused && !disabled
                                ? focusedRenderer({
                                      getInputProps: getInputPropsWithUpdatedRef(getInputProps),
                                      getToggleButtonProps,
                                      onBlur: handleBlur,
                                      onFocus: handleInputOnFocus(openMenu),
                                      highlightedIndex,
                                      inputValue,
                                  })
                                : blurredRenderer({
                                      getInputProps,
                                      getToggleButtonProps,
                                      onFocus: handleInputOnFocus(openMenu),
                                  })}
                            <List
                                {...getMenuProps({
                                    ...elem('list', stateAndProps),
                                    ref: listRef.current,
                                    isControlledNavigation: true,
                                })}
                            >
                                {focused && !disabled
                                    ? listRenderer({
                                          suggestionToString,
                                          suggestions,
                                          getItemProps,
                                          highlightedIndex,
                                          inputValue: inputValueRecall,
                                      })
                                    : null}
                            </List>
                        </FieldWrapper>
                    </div>
                )}
            </Downshift>
        </div>
    );
}

SelectBase.defaultProps = {
    showClearButton: false,
    keepExpandedAfterSelection: false,
    clearInputAfterSelection: false,
    highlightOnEmptyInput: true,
    clearTitle: '',
    selectOnTab: false,
    onInputValueChange: () => null,
};

SelectBase.displayName = 'SelectBase';
