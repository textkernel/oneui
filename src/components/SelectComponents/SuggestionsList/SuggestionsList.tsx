import * as React from 'react';
import { GetItemPropsOptions } from 'downshift';
import { bem } from '../../../utils/bem';
import { ContentPlaceholder } from '../../ContentPlaceholder';
import { Text, MarkedText } from '../../Text';
import { ListOptimizer, ListItem } from '../../List';
import { NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS } from '../../../constants';
import styles from './SuggestionsList.scss';

const { elem } = bem('SuggestionsList', styles);

export interface Props<S> {
    /** An array of objects that will be used to render the suggestions list. */
    suggestions: S[];
    /** if suggestions are still loading, i.e. display placeholders */
    isLoading?: boolean;
    /** Enable ListOptimizer component for decreasing render time */
    useOptimizeRender?: boolean;
    /** suggestionToString(suggestion) should return a string to be displayed in the UI. e.g.: suggestion => suggestion.name */
    suggestionToString: (suggestion: S) => string;
    /** suggestionToKey(suggestion) makes a key to be used for a suggestion item */
    suggestionToKey?: (suggestion: S) => string;
    /** render function for suggestion list item. If undefined `suggestionToString` will be used */
    suggestionItemRenderer?: (suggestion: S, index: number, array: S[]) => React.ReactNode;
    /** to be shown when no suggestions are available */
    noSuggestionsPlaceholder?: string;
    /** Defines if the first item of suggestions list is always visible */
    allowMixingSuggestionsAndLoading?: boolean;
    /** a function which gets props for the item in the list */
    getItemProps: (options: GetItemPropsOptions<S>) => object;
    /** index of the item from the list to be highlighted */
    highlightedIndex: number | null;
    /** input field value to be highlighted in the item from the list */
    inputValue: string;
    /** In some cases you need to pass disabled attributes to the top level 'li' od each item
     * e.g. in the context of downshift to disable keyboard navigation on these items
     */
    passDisabledToListItems?: boolean;
}

export function SuggestionsList<S>({
    suggestionToString,
    useOptimizeRender = false,
    suggestions,
    isLoading = false,
    allowMixingSuggestionsAndLoading = false,
    noSuggestionsPlaceholder = '',
    getItemProps,
    highlightedIndex,
    suggestionToKey,
    suggestionItemRenderer,
    inputValue = '',
    passDisabledToListItems = false,
}: Props<S>) {
    // eslint-disable-next-line react/display-name
    const renderItem = ({ key, index, style = {} }) => {
        const currentItem = suggestions[index];
        let disabled = false;
        if (typeof currentItem === 'object' && currentItem !== null && 'disabled' in currentItem) {
            // eslint-disable-next-line @typescript-eslint/dot-notation
            disabled = currentItem['disabled']; // TS only happy with this syntax
        }

        return (
            <ListItem
                key={key}
                style={style}
                {...getItemProps({
                    item: currentItem,
                    index,
                    disabled,
                })}
                isHighlighted={highlightedIndex === index}
                passDisabledToLi={passDisabledToListItems}
            >
                {suggestionItemRenderer ? (
                    suggestionItemRenderer(currentItem, index, suggestions)
                ) : (
                    <MarkedText marker={inputValue} inline>
                        {suggestionToString(currentItem)}
                    </MarkedText>
                )}
            </ListItem>
        );
    };

    // eslint-disable-next-line react/display-name
    const renderLoadingPlaceholders = () =>
        Array(NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS)
            .fill('')
            .map((el, i) => {
                return (
                    // eslint-disable-next-line react/no-array-index-key
                    <ListItem key={i}>
                        <div {...elem('loaderItem')} role="listitem">
                            <ContentPlaceholder />
                        </div>
                    </ListItem>
                );
            });

    // eslint-disable-next-line react/display-name
    const renderSuggestionsList = () => (
        <>
            {useOptimizeRender ? (
                <div {...elem('optimizerWrapper')}>
                    <ListOptimizer rowCount={suggestions.length}>{renderItem}</ListOptimizer>
                </div>
            ) : (
                <>
                    {suggestions.map((item, index) => {
                        const key = suggestionToKey
                            ? suggestionToKey(item)
                            : suggestionToString(item);
                        return renderItem({ key, index });
                    })}
                </>
            )}
        </>
    );

    if (isLoading) {
        if (!allowMixingSuggestionsAndLoading || !suggestions || !suggestions.length) {
            return <>{renderLoadingPlaceholders()}</>;
        }

        return (
            <>
                {renderSuggestionsList()}
                {renderLoadingPlaceholders()}
            </>
        );
    }

    if (!suggestions || !suggestions.length) {
        if (noSuggestionsPlaceholder) {
            return (
                <ListItem disabled>
                    <Text context="neutral">{noSuggestionsPlaceholder}</Text>
                </ListItem>
            );
        }
        return null;
    }

    return renderSuggestionsList();
}

SuggestionsList.displayName = 'SuggestionsList';
