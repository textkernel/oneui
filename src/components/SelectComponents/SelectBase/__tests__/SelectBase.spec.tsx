import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { SelectBase } from '../SelectBase';
import { SuggestionsList } from '../../SuggestionsList';
import {
    SUGGESTIONS,
    SUGGESTION_TO_STRING,
} from '../../../AutosuggestDeprecated/__mocks__/suggestions';

describe('SelectBase', () => {
    const inputRef = React.createRef();
    const suggestions = SUGGESTIONS;
    const suggestionToString = SUGGESTION_TO_STRING;
    const mockOnSelectionAdd = jest.fn();
    const mockOnInputValueChange = jest.fn();
    const mockOnClearAllSelected = jest.fn();
    const mockOnBlur = jest.fn();
    const mockRender = jest.fn(({ getInputProps }) => <input {...getInputProps()} />);
    const mockRenderWithRef = jest.fn(({ getInputProps }) => (
        <input ref={inputRef} {...getInputProps()} />
    ));

    let view;
    let inputNode;

    const defaultProps = {
        suggestions,
        suggestionToString,
        // eslint-disable-next-line react/display-name
        listRenderer: (listProps) => <SuggestionsList {...listProps} />,
        focusedRenderer: mockRender,
        blurredRenderer: mockRender,
        onSelectionAdd: mockOnSelectionAdd,
        onInputValueChange: mockOnInputValueChange,
        onClearAllSelected: mockOnClearAllSelected,
        onBlur: mockOnBlur,
        highlightOnEmptyInput: true,
    };

    const rerenderView = (props) => {
        view.rerender(<SelectBase {...defaultProps} {...props} />);
    };

    beforeEach(() => {
        view = render(<SelectBase {...defaultProps} />);
        inputNode = screen.getByRole('textbox');
    });

    describe('rendering', () => {
        it('should initially render empty component correctly', () => {
            expect(view.asFragment()).toMatchSnapshot();
        });
    });
    describe('with toggle arrow', () => {
        it('should show arrows when showArrow is true', () => {
            const newProps = {
                highlightOnEmptyInput: true,
                showArrow: true,
            };
            rerenderView(newProps);
            expect(view.asFragment()).toMatchSnapshot();

            expect(screen.getByRole('button')).toBeInTheDocument();
        });
        it('should toggle focus and the arrow when it is clicked', async () => {
            const newProps = {
                highlightOnEmptyInput: true,
                showArrow: true,
            };
            rerenderView(newProps);
            expect(view.asFragment()).toMatchSnapshot();
            const svg = screen.getByRole('button');

            expect(svg).toBeInTheDocument();
            await userEvent.click(svg);
            const svgFocused = screen.getByRole('button');
            expect(svgFocused).toBeInTheDocument();
            await userEvent.click(svgFocused);
            expect(svgFocused).not.toBeInTheDocument();
        });
    });
    describe('search field interactions', () => {
        it('should set focus when wrapper element is clicked', async () => {
            const focusSpy = jest.spyOn(inputNode, 'focus');

            expect(inputNode).not.toBe(document.activeElement);
            expect(focusSpy).not.toHaveBeenCalled();

            await userEvent.click(screen.queryAllByRole('listbox')[0]);

            expect(inputNode).toBe(document.activeElement);
            expect(focusSpy).toHaveBeenCalled();
        });
        it('should be able to get a component by ref sent as a prop', () => {
            const newProps = {
                ref: inputRef,
                focusedRenderer: mockRenderWithRef,
                blurredRenderer: mockRenderWithRef,
                highlightOnEmptyInput: true,
            };
            rerenderView(newProps);

            expect(screen.getAllByRole('textbox')[0]).toBeInTheDocument();
        });
        it('should lose focus when suggestion is selected', async () => {
            expect(inputNode).not.toBe(document.activeElement);

            await userEvent.click(screen.queryAllByRole('listbox')[0]);
            await userEvent.click(screen.queryAllByRole('presentation')[0]);

            expect(screen.queryAllByRole('presentation')).toHaveLength(0);
        });
        it('should stay focused when suggestion is selected with keepExpandedAfterSelection set to true', async () => {
            const newProps = {
                highlightOnEmptyInput: true,
                keepExpandedAfterSelection: true,
            };
            rerenderView(newProps);
            expect(inputNode).not.toBe(document.activeElement);

            await userEvent.click(screen.queryAllByRole('listbox')[0]);
            expect(inputNode).toBe(document.activeElement);

            await userEvent.click(screen.queryAllByRole('presentation')[0]);
            expect(screen.queryAllByRole('presentation')).toHaveLength(suggestions.length);
        });
        it('should clear the input field when a suggestion was selected', async () => {
            const textInputValue = 'driver';
            const inputField = screen.getAllByRole('textbox')[0];
            await userEvent.type(inputField, textInputValue);

            expect(inputField.getAttribute('value')).toEqual(textInputValue);

            await userEvent.click(screen.queryAllByRole('presentation')[0]);

            expect(inputField.getAttribute('value')).toEqual('');
        });
        it('should clear the input field when a suggestion was selected with keepExpandedAfterSelection set to true', async () => {
            const textInputValue = 'driver';
            const newProps = {
                highlightOnEmptyInput: true,
                keepExpandedAfterSelection: true,
            };
            rerenderView(newProps);
            const inputField = screen.getAllByRole('textbox')[0];
            await userEvent.type(inputField, textInputValue);

            expect(inputField.getAttribute('value')).toEqual(textInputValue);

            await userEvent.click(screen.queryAllByRole('presentation')[0]);

            expect(inputField.getAttribute('value')).toEqual(textInputValue);
        });
    });
    describe('highlighting', () => {
        it('should highlight first item', () => {
            expect(view.asFragment()).toMatchSnapshot();
            expect(screen.queryAllByRole('listbox')[0]).toBeInTheDocument();
        });
    });

    describe('callbacks', () => {
        describe('onSelectionAdd', () => {
            it('should be called on clicking on a suggestion', async () => {
                await userEvent.click(screen.queryAllByRole('listbox')[0]);

                expect(mockOnSelectionAdd).not.toHaveBeenCalled();

                await userEvent.click(screen.queryAllByRole('presentation')[0]);

                expect(mockOnSelectionAdd).toHaveBeenCalled();
            });
        });
        it('should call onClearAllSelected on Clear button click', async () => {
            const clearTitle = 'Clear';
            const newProps = {
                highlightOnEmptyInput: true,
                showArrow: true,
                clearTitle,
                showClearButton: true,
            };
            rerenderView(newProps);
            expect(view.asFragment()).toMatchSnapshot();
            const clearButton = screen.getByRole('button', { name: 'Clear' });
            await userEvent.click(clearButton);

            expect(mockOnClearAllSelected).toHaveBeenCalled();
        });
        // Despite everything is working good in real-case scenario,
        // 'blur' event can not be simulated in the way Downshift component to understand it.
        it.skip('should call onBlur', async () => {
            await userEvent.click(screen.queryAllByRole('listbox')[0]);

            await userEvent.click(screen.queryAllByRole('listbox')[0]);

            expect(mockOnBlur).toHaveBeenCalled();
        });
        it('should call onInputValueChange when typing into input field', async () => {
            expect(mockOnInputValueChange).not.toHaveBeenCalled();

            const inputField = screen.getAllByRole('textbox')[0];
            await userEvent.type(inputField, 'drive');

            expect(mockOnInputValueChange).toHaveBeenCalled();
        });
    });
});
