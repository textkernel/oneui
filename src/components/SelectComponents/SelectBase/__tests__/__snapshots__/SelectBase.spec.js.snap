// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBase rendering should initially render empty component correctly 1`] = `
<SelectBase
  blurredRenderer={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "getInputProps": [Function],
            "getToggleButtonProps": [Function],
            "onFocus": [Function],
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-labelledby="downshift-0-label"
            autoComplete="off"
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />,
        },
      ],
    }
  }
  clearInputAfterSelection={false}
  clearTitle=""
  focusedRenderer={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "getInputProps": [Function],
            "getToggleButtonProps": [Function],
            "onFocus": [Function],
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-labelledby="downshift-0-label"
            autoComplete="off"
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />,
        },
      ],
    }
  }
  highlightOnEmptyInput={true}
  keepExpandedAfterSelection={false}
  listRenderer={[Function]}
  onBlur={[MockFunction]}
  onClearAllSelected={[MockFunction]}
  onInputValueChange={[MockFunction]}
  onSelectionAdd={[MockFunction]}
  selectOnTab={false}
  showClearButton={false}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
>
  <div
    className="SelectBase"
  >
    <Downshift
      defaultHighlightedIndex={0}
      defaultIsOpen={false}
      environment={[Window]}
      getA11yStatusMessage={[Function]}
      inputValue=""
      itemToString={[Function]}
      onChange={[Function]}
      onInputValueChange={[Function]}
      onOuterClick={[Function]}
      onSelect={[Function]}
      onStateChange={[Function]}
      onUserAction={[Function]}
      scrollIntoView={[Function]}
      selectedItemChanged={[Function]}
      stateReducer={[Function]}
      suppressRefError={false}
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label"
        aria-owns={null}
        className="SelectBase__main"
        role="combobox"
      >
        <FieldWrapper
          className="SelectBase__field"
          clearLabel=""
          disabled={false}
          isFocused={false}
          onClear={[Function]}
          onClick={[Function]}
          showClearButton={false}
        >
          <div
            className="FieldWrapper SelectBase__field"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="FieldWrapper__content"
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-controls={null}
                aria-labelledby="downshift-0-label"
                autoComplete="off"
                id="downshift-0-input"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                value=""
              />
              <List
                aria-labelledby="downshift-0-label"
                doSelectOnNavigate={false}
                id="downshift-0-menu"
                isControlledNavigation={true}
                isDivided={false}
                role="listbox"
              >
                <ul
                  aria-labelledby="downshift-0-label"
                  className="List"
                  id="downshift-0-menu"
                  role="listbox"
                />
              </List>
            </div>
          </div>
        </FieldWrapper>
      </div>
    </Downshift>
  </div>
</SelectBase>
`;
